File: c/3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#define dataLine 33
#define defaultInput "../../Inputs/2020_3.txt"

int findLines(char *fileName) {
	int numLines = 0;
	char textRead[dataLine];

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	// Check if the file exists or not
    if (inFile == NULL) {
        return -1;
    }

	while(fgets(textRead, dataLine, inFile)) {
		++numLines;
	}

	fclose(inFile);

	return numLines;
}

int part1(char *fileName) {
	int numLines, count = 0;
	int slope[2] = {3, 1};
	int loc[2] = {0, 0};

	numLines = findLines(fileName);

	if (numLines == -1) {
		printf("Error: could not open file. Quitting\n");
		return -1;
	}

	char textRead[dataLine];
	int lineNum = 0;

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	while(fgets(textRead, dataLine, inFile)) {
        int charIndex = loc[0] % (dataLine - 2);
        if (textRead[charIndex] == '#') {
            count++;
            textRead[charIndex] = 'X';
        } else {
            textRead[charIndex] = 'O';
        }

        loc[0] += slope[0];
        loc[1] += slope[1];
	}

	fclose(inFile);

    return count;
}

int part2(char *fileName) {
	int numLines, count, product = 1;
	int slope[2];
	int loc[2];


	numLines = findLines(fileName);

	if (numLines == -1) {
		printf("Error: could not open file. Quitting\n");
		return -1;
	}

    for (int i = 0; i < 5; i++) {
        if (i == 0) {
            slope[0] = 1;
            slope[1] = 1;
        } else if (i == 4) {
            slope[0] = 1;
            slope[1] = 2;
        } else {
            slope[0] += 2;
        }

        loc[0] = 0;
        loc[1] = 0;

        count = 0;

        char textRead[dataLine];
        int lineNum = -1;

        // Open the file
        FILE *inFile = fopen(fileName, "r");

        while(fgets(textRead, dataLine, inFile)) {
            lineNum++;
            if (lineNum != loc[1]) {
                continue;
            }
            int charIndex = loc[0] % (dataLine - 2);
            if (textRead[charIndex] == '#') {
                count++;
                textRead[charIndex] = 'X';
            } else {
                textRead[charIndex] = 'O';
            }

            loc[0] += slope[0];
            loc[1] += slope[1];
        }

        fclose(inFile);

        // printf("Trees Hit: %d\n", count);

        product *= count;
    }

    return product;
}

int main (int argc, char *argv[]) {
    char *inputPath = defaultInput;
    if (argc > 1) {
        inputPath = argv[1];
    }

    clock_t t; 
    t = clock(); 
    int p1 = part1(inputPath);
    t = clock() - t; 
    double t_p1 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 1:\nTrees hit: %d\nRan in %f seconds\n", p1, t_p1);

    t = clock();
    int p2 = part2(inputPath);
    t = clock() - t;
    double t_p2 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 2:\nProduct of trees hit at different slopes: %d\nRan in %f seconds\n", p2, t_p2);

	return 0;
}

File: python/3.py
import re
import sys
from pathlib import Path
from typing import Any, List, Optional, Tuple

sys.path.append(str(Path(__file__).parent.parent.parent))
from Modules.timer import Timer
def part1(data):
    """ 2020 Day 3 Part 1

    >>> part1(['..##.......', '#...#...#..', '.#....#..#.', '..#.#...#.#', '.#...##..#.', '..#.##.....', '.#.#.#....#', '.#........#', '#.##...#...', '#...##....#', '.#..#...#.#'])
    7
    """

    x, y = 0, 0
    dx, dy = 3, 1

    count = 0
    while y < len(data):
        if data[y][x] == '#':
            count += 1

        x += dx
        x %= len(data[0])
        y += dy

    return count


def part2(data):
    """ 2020 Day 3 Part 2

    >>> part2(['..##.......', '#...#...#..', '.#....#..#.', '..#.#...#.#', '.#...##..#.', '..#.##.....', '.#.#.#....#', '.#........#', '#.##...#...', '#...##....#', '.#..#...#.#'])
    336
    """

    product = part1(data)
    for dx, dy in [(1, 1), (5, 1), (7, 1), (1, 2)]:
        x, y = 0, 0

        count = 0
        while y < len(data):
            if data[y][x] == '#':
                count += 1

            x += dx
            x %= len(data[0])
            y += dy

        product *= count

    return product


def main(input_path: Optional[Path | str]=None, verbose: bool=False) -> Tuple[Tuple[Any, float]]:
    if not input_path:
        if not (input_path := sys.argv[1] if len(sys.argv) > 1 else None):
            year, day = re.findall(r'\d+', str(__file__))[-2:]
            input_path = Path(Path(__file__).parent.parent.parent, "Inputs", f"{year}_{day}.txt")
    
    with open(input_path, encoding='UTF-8') as f:
        data = [line.strip('\n') for line in f.readlines()]

    with Timer() as p1_time:
        p1 = part1(data)

    if verbose:
        print(f"\nPart 1:\nTrees hit: {p1}\nRan in {p1_time.elapsed:0.4f} seconds")

    with Timer() as p2_time:
        p2 = part2(data)

    if verbose:
        print(f"\nPart 2:\nProduct of trees hit at different slopes: {p2}\nRan in {p2_time.elapsed:0.4f} seconds")

    return [(p1, p1_time.elapsed), (p2, p2_time.elapsed)]


if __name__ == "__main__":
    main(verbose=True)